### PR TITLE
Make DifficultyIcon support dynamic star difficulty

### PR DESCRIPTION
**Pull request by smoogipoo**
Preview: https://drive.google.com/file/d/1syCxOYoG6ci56pWy_jUZZJxzGR30CgBi/view?usp=sharing

This is set up to work for timeshift, however it doesn't work in practice since timeshift doesn't retrieve the locally-available beatmap so it falls back to whatever's coming in via `BeatmapInfo.StarDifficulty`. _Something_ needs to retrieve the local beatmap somewhere, unsure if multiplayer or `BeatmapDifficultyManager` at this point. I think there was a reason why I didn't do it for multiplayer.

This also doesn't support grouped difficulty icons yet.

The calculation is implemented through a DLUW to prevent offscreen calculations from taking place in e.g. song select.

### CODE REVIEW

**Comment by Gonzalo-Bruna** [comment]
is it a known issue that the star rating in the beatmap wedge doesn't update when you apply mods?

**Comment by bdach** [comment]
I'm pretty sure it is, the buffered container there doesn't update entirely correctly. It's a separate issue.

File: osu.Game/Beatmaps/BeatmapDifficultyManager.cs [hunk 1]
```diff
@@ -307,5 +307,19 @@ public StarDifficulty(double stars, int maxCombo)
 
             // Todo: Add more members (BeatmapInfo.DifficultyRating? Attributes? Etc...)
         }
+
+        public DifficultyRating DifficultyRating
+        {
+            get
+            {
+                if (Stars < 2.0) return DifficultyRating.Easy;
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapDifficultyManager.cs:315]: where are these values coming from? osu-web? may be good to leave an inline comment linking it up.

File: osu.Game/Beatmaps/Drawables/DifficultyIcon.cs [hunk 2]
```diff
@@ -82,16 +110,73 @@ private void load(OsuColour colours)
...
             };
+
+            difficultyBindable.BindValueChanged(difficulty => background.Colour = colours.ForDifficultyRating(difficulty.NewValue.DifficultyRating));
+        }
+
+        public ITooltip GetCustomTooltip() => new DifficultyIconTooltip();
+
+        public object TooltipContent => shouldShowTooltip ? new DifficultyIconTooltipContent(beatmap, difficultyBindable) : null;
+
+        private class DifficultyRetriever : Drawable
```
Comment (peppy) [inline osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:125]:
```suggestion
        private class DifficultyRetriever : Component
```

not hugely important, but better defines what it is.

**Review by peppy — CHANGES_REQUESTED** [review]

**Review by bdach — APPROVED** [review]
this looks real cool :+1:

**Review by peppy — APPROVED** [review]

**Pull request merged**